src/qibo/transpiler/router.py

```diff
@@ -875,6 +897,42 @@ def _execute_blocks(self, blocklist: list):
         self._update_front_layer()
         self._memory_map = []
         self._delta_register = [1.0 for _ in self._delta_register]
+        self._temporary_added_swaps = 0
+        self._saved_circuit = deepcopy(self.circuit)
+
+    def _shortest_path_routing(self):
+        """Route a gate in the front layer using the shortest path. This method is executed when the standard SABRE fails to find an optimized solution.
+
+        Method works in-place.
+        """
+
+        min_distance = float("inf")
+        shortest_path_qubits = None
+
+        for block in self._front_layer:
+            q1 = self.circuit.get_physical_qubits(block)[0]
+            q2 = self.circuit.get_physical_qubits(block)[1]
+            distance = self._dist_matrix[q1, q2]
+
+            if distance < min_distance:
+                min_distance = distance
+                shortest_path_qubits = [q1, q2]
+
+        shortest_path = nx.bidirectional_shortest_path(
+            self.connectivity, shortest_path_qubits[0], shortest_path_qubits[1]
+        )
+
+        # Q1 is moved
+        swaps = [
+            (
+                self.circuit.physical_to_logical(shortest_path[i]),
+                self.circuit.physical_to_logical(shortest_path[i + 1]),
+            )
+            for i in range(len(shortest_path) - 2)
+        ]
```

Comment:
Thank you for your suggestion, but I think your suggestion differs from the original code in two ways.

- Since it is enough to move q1 adjacent to q2, it doesn't require a swap between q1 and q2. So it would be something like:

```py
... shortest_path[0:-2], shortest_path[1: -1] ...
```

- Since the `shortest_path` array contains physical qubit numbers and the swap occurs on logical qubit numbers, `physical_to_logical` should be applied to all elements of `shortest_path`.

Therefore, the equivalent code would be:

```py
swaps = [
    (
        self.circuit.physical_to_logical(a), 
        self.circuit.physical_to_logical(b)
    ) 
    for a, b in zip(shortest_path[0:-2], shortest_path[1:-1])
]
```

Is it better to change the code to the one above?

Reply:
Ah yeah sorry, then better to first convert with `physical_to_logical` and then build the pairs in order to avoid converting the same element twice:
``` python
shortest_path = [self.circuit.physical_to_logical(q) for q in shortest_path[:-1]]
swaps = zip(shortest_path[:-1], shortest_path[1:])
```